lib/Secrets/secrets_c.c

```diff
@@ -0,0 +1,1078 @@
+        fprintf(stderr, "Message has no arguments!\n");
+    } else if (DBUS_TYPE_STRING != dbus_message_iter_get_arg_type(&args)) {
+        fprintf(stderr, "Argument is not string!\n");
+    } else {
+        dbus_message_iter_get_basic(&args, &param);
+    }
+
+    *p2pAddr = strdup(param);
+
+    dbus_error_free(&err);
+    dbus_message_unref(msg);
+
+    return 1;
+}
+
+static struct Sailfish_Secrets_DBus_Connection {
+    DBusConnection *sessionBus;
+    DBusConnection* p2pBus;
+    DBusMessage* msg;
+    DBusPendingCall* pending;
+    char *p2pAddr;
+} daemon_connection = {
+    .sessionBus = NULL,
+    .p2pBus = NULL,
+    .msg = NULL,
+    .pending = NULL,
+    .p2pAddr = NULL
+};
+
+int Sailfish_Secrets__isConnectedToServer()
+{
+    return daemon_connection.p2pBus == NULL ? 0 : 1;
+}
+
+int Sailfish_Secrets__connectToServer()
+{
+    DBusError daemon_connection_error;
+
+    if (daemon_connection.p2pBus != NULL) {
+        fprintf(stderr, "Already connected to secrets daemon!\n");
+        return 1; /* Return "ok" */
+    }
+
+    /* Get the peer to peer address of the daemon via the session bus */
+    if (!daemon_connection.p2pAddr &&
+            !Sailfish_Secrets__getServerPeerToPeerAddress(
+                &daemon_connection.p2pAddr,
+                &daemon_connection.sessionBus)) {
+        return 0;
+    }
+
+    /* Open a private peer to peer connection to the daemon */
+    dbus_error_init(&daemon_connection_error);
+    daemon_connection.p2pBus = dbus_connection_open_private(
+                daemon_connection.p2pAddr,
+                &daemon_connection_error);
+    if (dbus_error_is_set(&daemon_connection_error)) {
+        fprintf(stderr,
+                "Connection Error (%s) to: %s\n",
+                daemon_connection_error.message,
+                daemon_connection.p2pAddr);
+        dbus_error_free(&daemon_connection_error);
+    }
+    if (daemon_connection.p2pBus == NULL) {
+        return 0;
+    }
+
+    return 1;
+}
+
+void Sailfish_Secrets__disconnectFromServer()
+{
+    dbus_connection_close(daemon_connection.p2pBus);
+    dbus_connection_unref(daemon_connection.p2pBus);
+    daemon_connection.p2pBus = NULL;
+    free(daemon_connection.p2pAddr);
+    daemon_connection.p2pAddr = NULL;
+}
+
+void Sailfish_Secrets__appendEnumArg(DBusMessageIter *in_args,
+                                    DBusMessageIter *enum_struct,
+                                    int *arg)
+{
+    dbus_message_iter_open_container(
+                in_args,
+                DBUS_TYPE_STRUCT,
+                NULL,
+                enum_struct);
+    dbus_message_iter_append_basic(
+                enum_struct,
+                DBUS_TYPE_INT32,
+                arg);
+    dbus_message_iter_close_container(
+                in_args,
+                enum_struct);
+}
+
+void Sailfish_Secrets__appendSecretIdentifierArg(
+        DBusMessageIter *args,
+        struct Sailfish_Secrets_Secret_Identifier *ident)
```

Comment:
One may want to factorise code here between secret and crypto. The identifiers are the same struct of two char* fields, they maybe shared and DBUS code handling these can be shared too.

Reply:
This was an explicit decision, to keep the two libraries entirely separate (that is, clients which link to libsailfishsecrets do not have to link to libsailfishcrypto and vice versa).  I guess we could use some private implementation code sharing inclusion tricks to share implementation but for such a small struct, I don't see the point, really.